spin_coater.c

```diff
@@ -0,0 +1,205 @@
+    count_of_measurements++;
+    return;
+  }
+
+  if (count_of_measurements < NUMBER_OF_HOLES) {
+    count_of_measurements++;
+    return;
+  }
+
+  const uint32_t current_time_us = time_us_32();
+  const uint32_t time_elapsed_us = current_time_us - last_time;
+  if((MICROSECONDS_PER_MINUTE / time_elapsed_us) < rpm_counter_based_on_photo_sensor + 1000)
+    rpm_counter_based_on_photo_sensor = (MICROSECONDS_PER_MINUTE / time_elapsed_us);
+  count_of_measurements = 0;
+}
+#endif
+
+#if  CONFIG_SPIN_COATER == CONFIG_SPIN_COATER_DSHOT || CONFIG_SPIN_COATER == CONFIG_SPIN_COATER_DSHOT_WITH_TELEMETRY
+int
+set_dshot_safe(furnace_context_t *ctx, unsigned dshot)
+{
+  if (dshot < SPIN_COATER_MIN_THROTTLE_COMMAND || dshot > SPIN_COATER_MAX_THROTTLE_COMMAND)
+    return 1;
+
+  ctx->spin_coater.dshot_throttle_val = dshot;
+  DEBUG_printf("\n DSHOT set to: %lu!\n", dshot);
+  dshot_send_command(ctx->spin_coater.dshot_throttle_val);
+  return 0;
+}
+#elif CONFIG_SPIN_COATER == CONFIG_SPIN_COATER_PWM
+int
+set_spin_coater_pwm_safe(furnace_context_t* ctx, unsigned duty)
+{
+  if (duty < SPIN_COATER_PWM_490_FREQ_DUTY_MIN || duty > SPIN_COATER_PWM_490_FREQ_DUTY_MAX)
+    return 1;
+
+  ctx->spin_coater.pwm_duty = duty;
+  DEBUG_printf("\n PWM set to: %lu!\n", duty);
+  return pwm_set_freq_duty(ctx->spin_coater.pwm_slice_num, PWM_CHAN_A, SPIN_COATER_PWM_490_FREQ, duty);
+}
+#endif
+// TODO refactor this code to support PWM smooth transition in future, right now it's not that important
+// cause PWM is not very usefull
+// We can consider dropping support for PWM in future.
+#if CONFIG_SPIN_COATER == CONFIG_SPIN_COATER_DSHOT || CONFIG_SPIN_COATER == CONFIG_SPIN_COATER_DSHOT_WITH_TELEMETRY
+static inline int
+scale_dshot_value_when_speeding(furnace_context_t* ctx, int rpm_left, int bias)
+{
+    return (15.0 / ctx->spin_coater.set_rpm) * rpm_left + (bias%2);
+}
+
+static inline int
+scale_dshot_value_when_stopping(furnace_context_t* ctx, int rpm_left)
+{
+    return 15.0 - (15.0 / ctx->spin_coater.set_rpm) * rpm_left;
+}
+
+static absolute_time_t
+do_dshot_smooth_transition(furnace_context_t* ctx)
+{
+#if CONFIG_SPIN_COATER == CONFIG_SPIN_COATER_DSHOT_WITH_TELEMETRY
+  ctx->spin_coater.current_rpm = calc_avg_from_raw_dshot_buffer(&ctx->spin_coater);
+#else 
+  ctx->spin_coater.current_rpm = rpm_counter_based_on_photo_sensor;
+#endif
+  int direction = ctx->spin_coater.current_rpm < ctx->spin_coater.set_rpm ? 1 : -1;
+  int rpm_diff_abs = abs(ctx->spin_coater.set_rpm - ctx->spin_coater.current_rpm);
+  absolute_time_t next_delay;
+  if(SPIN_STARTED_WITH_TIMER == ctx->spin_coater.spin_state) {
+    if(rpm_diff_abs > 200){
+      ctx->spin_coater.dshot_throttle_val += scale_dshot_value_when_speeding(ctx,
+                                                                            rpm_diff_abs,
+                                                                            (ctx->spin_coater.current_rpm/(double)(ctx->spin_coater.set_rpm)))*direction;
```

Comment:
Can we split this into multiple intermediate variables and make the lines shorter?

Reply:
removed it. Looks like it doesn't have any impact currently.